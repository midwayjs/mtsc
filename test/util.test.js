const { parseArgs, deleteFolderRecursive, copyFilesRecursive, readJSONCFile, filterFileChangedText } = require('../lib/util');
const fs = require('fs');
const path = require('path')

describe('test/util.test.js', () => {
  it('should parse', () => {
    expect(parseArgs(['node', 'tsc'])).toEqual([undefined, [], [], false]);
    expect(parseArgs(['node', 'tsc', '--watch'])).toEqual([undefined, ['--watch'], [], false]);
    expect(parseArgs(['node', 'tsc', '--watch', '--run', 'index.js'])).toEqual(['index.js', ['--watch'], [], false]);
    expect(parseArgs(['node', 'tsc', '--watch', '--run', 'index.js', '--', '--port', '3000'])).toEqual(['index.js', ['--watch'], ['--', '--port', '3000'], false]);
    expect(parseArgs(['node', 'tsc', '--watch', '--project', 'tsconfig.prod.json'])).toEqual([undefined, ['--watch', '--project', 'tsconfig.prod.json'], [], false]);
    expect(parseArgs(['node', 'tsc', '--watch', '--run'])).toEqual([undefined, ['--watch'], [], false]);
    expect(parseArgs(['node', 'tsc', '--cleanOutDir'])).toEqual([undefined, [], [], true]);
    expect(parseArgs(['node', 'tsc', '--watch', '--cleanOutDir', '--build', '--run'])).toEqual([undefined, ['--watch', '--build'], [], true]);
  });

  describe('deleteFolderRecursive', () => {
    const testDirPath = path.join(__dirname, 'testDir');

    // 在每个测试用例之前，创建一个测试目录和一些测试文件
    beforeEach(() => {
      fs.mkdirSync(testDirPath, { recursive: true });
      fs.writeFileSync(path.join(testDirPath, 'file1.txt'), 'Hello, world!');
      fs.mkdirSync(path.join(testDirPath, 'subDir'));
      fs.writeFileSync(path.join(testDirPath, 'subDir', 'file2.txt'), 'Hello, world!');
    });

    // 在每个测试用例之后，确保测试目录已被删除
    afterEach(() => {
      if (fs.existsSync(testDirPath)) {
        deleteFolderRecursive(testDirPath);
      }
    });

    it('should delete the directory and all its contents', () => {
      deleteFolderRecursive(testDirPath);

      // 检查目录是否已被删除
      expect(fs.existsSync(testDirPath)).toBe(false);
    });
  });


  describe('copyFilesRecursive', () => {
    const sourceDir = path.join(__dirname, 'fixtures/copy/src');
    const targetDir = path.join(__dirname, 'targetDir/dist');

    // 在每个测试用例之后，确保测试目录已被删除
    beforeEach(() => {
      if (fs.existsSync(targetDir)) {
        deleteFolderRecursive(targetDir);
      }
    });

    it('should copy the directory and all its contents', () => {
      copyFilesRecursive(sourceDir, targetDir, true);

      // 检查目录是否已被复制
      expect(fs.existsSync(targetDir)).toBe(true);
      expect(fs.existsSync(path.join(targetDir, 'a.ts'))).toBe(false);
      expect(fs.existsSync(path.join(targetDir, 'a.js'))).toBe(false);
      expect(fs.existsSync(path.join(targetDir, 'b/.ccc'))).toBe(true);
      expect(fs.existsSync(path.join(targetDir, 'b/b.html'))).toBe(true);
    });

    it('should copy the directory and all its contents', () => {
      copyFilesRecursive(sourceDir, targetDir, false);

      // 检查目录是否已被复制
      expect(fs.existsSync(targetDir)).toBe(true);
      expect(fs.existsSync(path.join(targetDir, 'a.ts'))).toBe(false);
      expect(fs.existsSync(path.join(targetDir, 'a.js'))).toBe(true);
      expect(fs.existsSync(path.join(targetDir, 'b/.ccc'))).toBe(true);
      expect(fs.existsSync(path.join(targetDir, 'b/b.html'))).toBe(true);
    });
  });


  it('should test read tsconfig file', () => {
    const data = readJSONCFile(path.join(__dirname, 'fixtures/ts_config/tsconfig.json'));
    expect(data.options).toBeDefined();
    expect(data.raw).toBeDefined();
  });

  it('should test read tsconfig file with extends', () => {
    const data = readJSONCFile(path.join(__dirname, 'fixtures/ts_config/tsconfig_extend.json'));
    expect(data.options).toBeDefined();
    expect(data.options.paths).toBeDefined();
    expect(data.raw).toBeDefined();
    expect(data.raw.paths).not.toBeDefined();
  });

  it('should parse output TSFILE', () => {
    expect(filterFileChangedText('TSFILE: /Users/harry/project/application/open-koa-v3/dist/task/hello.d.ts')).toEqual(['', []]);
    expect(filterFileChangedText('TSFILE: /Users/harry/project/application/open-koa-v3/dist/task/hello.js')).toEqual(['', ['/Users/harry/project/application/open-koa-v3/dist/task/hello.js']]);
  });

});
