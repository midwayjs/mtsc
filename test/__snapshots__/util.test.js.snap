// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/util.test.js should test read tsconfig file 1`] = `
{
  "compileOnSave": true,
  "errors": [
    {
      "category": 1,
      "code": 18003,
      "file": undefined,
      "length": undefined,
      "messageText": "No inputs were found in config file 'tsconfig.json'. Specified 'include' paths were '["**/*"]' and 'exclude' paths were '["dist","node_modules","test"]'.",
      "reportsDeprecated": undefined,
      "reportsUnnecessary": undefined,
      "start": undefined,
    },
  ],
  "fileNames": [],
  "options": {
    "configFilePath": undefined,
    "module": 1,
    "moduleResolution": 2,
    "target": 5,
  },
  "projectReferences": undefined,
  "raw": {
    "compileOnSave": true,
    "compilerOptions": {
      "module": "commonjs",
      "moduleResolution": "node",
      "target": "es2018",
    },
    "exclude": [
      "dist",
      "node_modules",
      "test",
    ],
  },
  "typeAcquisition": {
    "enable": false,
    "exclude": [],
    "include": [],
  },
  "watchOptions": undefined,
  "wildcardDirectories": {
    "/users/harry/project/mtsc/test/fixtures/ts_config": 1,
  },
}
`;

exports[`test/util.test.js should test read tsconfig file with extends 1`] = `
{
  "compileOnSave": false,
  "errors": [
    {
      "category": 1,
      "code": 18003,
      "file": undefined,
      "length": undefined,
      "messageText": "No inputs were found in config file 'tsconfig.json'. Specified 'include' paths were '["src/**/*.ts"]' and 'exclude' paths were '["asset","app/public","app/views","dist","node_modules*","**/*.d.ts","**/*.spec.ts"]'.",
      "reportsDeprecated": undefined,
      "reportsUnnecessary": undefined,
      "start": undefined,
    },
  ],
  "fileNames": [],
  "options": {
    "baseUrl": "/Users/harry/project/mtsc/test/fixtures/ts_config",
    "configFilePath": undefined,
    "declaration": true,
    "declarationMap": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "incremental": true,
    "module": 199,
    "moduleResolution": 99,
    "newLine": 1,
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "/Users/harry/project/mtsc/test/fixtures/ts_config/dist",
    "paths": {
      "##/*": [
        "src/*",
      ],
      "#@/*": [
        "test/*",
      ],
      "#a/*": [
        "src/app/*",
      ],
    },
    "pathsBasePath": "/Users/harry/project/mtsc/test/fixtures/ts_config",
    "preserveConstEnums": true,
    "pretty": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "strictPropertyInitialization": false,
    "stripInternal": true,
    "target": 9,
    "tsBuildInfoFile": "/Users/harry/project/mtsc/test/fixtures/ts_config/.tsbuildinfo",
    "types": [
      "mocha",
      "node",
    ],
  },
  "projectReferences": undefined,
  "raw": {
    "compileOnSave": false,
    "compilerOptions": {
      "paths": {
        "##/*": [
          "src/*",
        ],
        "#@/*": [
          "test/*",
        ],
        "#a/*": [
          "src/app/*",
        ],
      },
    },
    "exclude": [
      "asset",
      "app/public",
      "app/views",
      "dist",
      "node_modules*",
      "**/*.d.ts",
      "**/*.spec.ts",
    ],
    "extends": "./tsconfig.base.json",
    "include": [
      "src/**/*.ts",
    ],
  },
  "typeAcquisition": {
    "enable": false,
    "exclude": [],
    "include": [],
  },
  "watchOptions": undefined,
  "wildcardDirectories": {
    "/users/harry/project/mtsc/test/fixtures/ts_config/src": 1,
  },
}
`;
